vm/systemSmartContracts/governance.go

```diff
@@ -288,521 +499,599 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
+	}
+
+	voteData.UsedPower.Add(voteData.UsedPower, vote.Power)
+	voteData.VoteItems = append(voteData.VoteItems, vote)
+
+	return voteData, proposal, nil
 }
 
-func (g *governanceContract) voteForProposal(
-	proposal []byte,
-	vote string,
-	voter []byte,
-	numVotes int32,
-) error {
-	voteData, err := g.getOrCreateVoteData(proposal, voter)
+// saveNewVoteData first saves the main vote data of the voter, then the full proposal with the updated information
+func (g *governanceContract) saveNewVoteData(voter []byte, voteData *VoteData, proposal *GeneralProposal, voteType VoteType) error {
+	proposalKey := append([]byte(proposalPrefix), proposal.GitHubCommit...)
+	voteItemKey := append(proposalKey, voter...)
+
+	marshaledVoteItem, err := g.marshalizer.Marshal(voteData)
 	if err != nil {
-		log.Warn("getOrCreateVoteData", "err", err)
 		return err
 	}
-	if voteData.NumVotes == numVotes && voteData.VoteValue == vote {
-		return nil
+	g.eei.SetStorage(voteItemKey, marshaledVoteItem)
+
+	if !g.proposalContainsVoter(proposal, voteItemKey) {
+		proposal.Votes = append(proposal.Votes, voteItemKey)
 	}
 
-	oldNum := voteData.NumVotes
-	oldValue := voteData.VoteValue
+	marshaledProposal, err := g.marshalizer.Marshal(proposal)
+	g.eei.SetStorage(proposalKey, marshaledProposal)
 
-	voteData.NumVotes = numVotes
-	voteData.VoteValue = vote
-	err = g.saveVoteValue(proposal, voter, voteData)
+	err = g.setLock(voter, voteType, proposal)
 	if err != nil {
-		log.Warn("saveVoteValue", "err", err)
-		return err
+		return nil
 	}
 
-	generalProposal, err := g.getGeneralProposal(proposal)
+	return nil
+}
+
+// saveVoteData saves the provided vote data into the storage
+func (g *governanceContract) saveVoteData(voter []byte, voteData *VoteData, proposalReference []byte) error {
+	proposalKey := append([]byte(proposalPrefix), proposalReference...)
+	voteItemKey := append(proposalKey, voter...)
+
+	marshaledVoteItem, err := g.marshalizer.Marshal(voteData)
 	if err != nil {
 		return err
 	}
-	currentNonce := g.eei.BlockChainHook().CurrentNonce()
-	if currentNonce < generalProposal.StartVoteNonce {
-		return vm.ErrVotedForAProposalThatNotBeginsYet
-	}
+	g.eei.SetStorage(voteItemKey, marshaledVoteItem)
 
-	if currentNonce > generalProposal.EndVoteNonce {
-		return vm.ErrVotedForAnExpiredProposal
+	return nil
+}
+
+// setLock will set a storage key with the nonce until the funds for a specific voter are locked
+func (g *governanceContract) setLock(voter []byte, voteType VoteType, proposal *GeneralProposal) error {
+	prefix := []byte(validatorLockPrefix)
+	if voteType == Account {
+		prefix = append([]byte(accountLockPrefix), proposal.GitHubCommit...)
 	}
+	lockKey := append(prefix, voter...)
 
-	generalProposal.Voters = append(generalProposal.Voters, voter)
-	g.addVotedDataToProposal(generalProposal, oldValue, -oldNum)
-	g.addVotedDataToProposal(generalProposal, vote, numVotes)
+	proposalDuration := proposal.EndVoteNonce - proposal.StartVoteNonce
```

Comment:
This was exactly the intention. I mentioned in the document, it would otherwise be unfair, you could always vote in the last minute and never have any skin in the game for your vote. So the solution was for your funds to be locked for a period equal to the proposal duration.

Reply:
Ok, understood.